packages/block-editor/src/components/block-list/index.js

```diff
@@ -61,44 +56,66 @@ function getDeepestNode( node, type ) {
 	return node;
 }
 
-class BlockList extends Component {
-	constructor( props ) {
-		super( props );
-
-		this.onSelectionStart = this.onSelectionStart.bind( this );
-		this.onSelectionEnd = this.onSelectionEnd.bind( this );
-		this.setSelection = this.setSelection.bind( this );
-		this.updateNativeSelection = this.updateNativeSelection.bind( this );
-
-		this.ref = createRef();
-	}
-
-	componentDidUpdate() {
-		this.updateNativeSelection();
-	}
+function BlockList( {
+	className,
+	rootClientId,
+	__experimentalMoverDirection: moverDirection = 'vertical',
+	isDraggable,
+	renderAppender,
+} ) {
+	const selector = ( select ) => {
```

Comment:
Nit: First time I see this pattern (assigning a variable instead of just using an inline function). Not against it but there's something to be said about consistency across our components.

Reply:
Agreed, I think it would make more sense to just use a function declaration.

Reply:
Then the constant names will clash and I’d have to rename them

Reply:
In other words, I believe this is needed because some selectors have the same name as the constants, for example `isSelectionEnabled: isSelectionEnabled()`.

Reply:
Assigning the constants on the next line fixes this issue.

Reply:
That's a good point. I wonder if should make it a guideline for all useSelect usage? cc @gziolo @aduth

Reply:
Hm, it doesn't feel quite as expressive to me as compared to the single statement usage, but I can appreciate the challenge you're trying to deal with here (I've encountered it myself as well).

A brain dump of assorted thoughts:

- With selectors being prefixed by some action verb, I expect most assignments shouldn't share the same name (i.e. the noun subject of the verb is typically used as the assigned name, `getBlocks` -> `blocks`). The main exception I see here is the "is" or "has" (boolean) prefixes. Not sure if we could have some different convention for how we name these variables as an alternative workaround.
- If my understanding of scoping is correct, technically this doesn't solve the problem that the shadowed variable names are in a shared scope (but the variables outside the `select` are uninitialized and cannot be accessed).
- Maybe we permit variable shadowing, at least in this specific usage. Ideally in a way that we don't need to use inline rule configuration.
- If we do decide to define the selector outside the argument in `useSelect`, could it then also make sense to define it outside the component altogether (i.e. more similar to what you did with `stopPropagation` [in this example](https://github.com/WordPress/gutenberg/pull/18813/files#diff-45808b84d474a0d4eb13cf6d3d2943aaR6))? Helps with the "still shared scope" point above, maybe potential optimization opportunity by the static reference to the selector function? Also potentially more intuitive to document than if it were inlined in the component, if that's something we want to do, if even just for proper types on the return values related to efforts at #18838.
- The more we name our functions, the better, since "(anonymous)" in a debugger call stack makes debugging our code more difficult than it needs to be.

Reply:
> If we do decide to define the selector outside the argument in useSelect, could it then also make sense to define it outside the component altogether

This won't always work as some of these use props or other component variables.

Reply:
> This won't always work as some of these use props or other component variables.

Ah yes, you're right. Better to have a consistent approach which works in all cases.

Reply:
@aduth @youknowriad  Exactly, I would encourage it if it's not dependent on component props, but it is the case it is.

Reply:
I changed it to a named function now.

Reply:
Anything I can change here for this PR? Should I renamed the constants or leave it like this with the selector function in front of the assignments?

Reply:
It seems fine to me as currently implemented. I think the bigger challenge will be to consider the impact for other use of `useSelect`, documenting these conventions if appropriate, etc.

Reply:
Would it work as:

```js
const result = useSelect( ( select ) => {
   //...all the logic goes here
} );
const { x, y, z } = result;
```

I don't know how the React linter for hooks work, but we should ensure that it catches the case when props are used in `useSelect` and they aren't passed as 2nd param:

```js
const result = useSelect( ( select ) => {
   //...all the logic goes here
   return {
       x: select( 'store' ).methodA( props.a ),
       x: select( 'store' ).methodB( props.b ),
       x: select( 'store' ).methodC( props.c ),
   };
}, [ props.a, props.b, props.c ] );
const { x, y, z } = result;
```

Well, it's probably something that we don't support at the moment anyway based on the code snippets I found. Anyway, I'm just sharing as something which might make the difference but doesn't have to :)

Reply:
That works too. I don't really mind either way. What shall we do?

Reply:
Let's leave it as it is and change these lines if we decide to do it differently?